src/behaviours/navigate-to-extension-page.test.tsx

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+import { getRendererExtensionFake, TestExtension } from "../renderer/components/test-utils/get-renderer-extension-fake";
+import React from "react";
+import { fireEvent, RenderResult } from "@testing-library/react";
+import isEmpty from "lodash/isEmpty";
+import queryParametersInjectable from "../renderer/routes/query-parameters.injectable";
+import currentPathInjectable from "../renderer/routes/current-path.injectable";
+import type { IComputedValue } from "mobx";
+import type { LensRendererExtension } from "../extensions/lens-renderer-extension";
+import { getApplicationBuilder } from "../renderer/components/test-utils/get-application-builder";
+
+describe("navigate to extension page", () => {
+  let rendered: RenderResult;
+  let testExtension: TestExtension;
+  let queryParameters: IComputedValue<object>;
+  let currentPath: IComputedValue<string>;
+
+  beforeEach(async () => {
+    const applicationBuilder = getApplicationBuilder();
+
+    testExtension = getRendererExtensionFake(
+      extensionWithPagesHavingParameters,
+    );
+
+    await applicationBuilder.addExtensions(testExtension);
+
+    rendered = await applicationBuilder.render();
+
+    const rendererDi = applicationBuilder.dis.rendererDi;
+
+    queryParameters = rendererDi.inject(queryParametersInjectable);
+    currentPath = rendererDi.inject(currentPathInjectable);
+  });
+
+  it("renders", () => {
+    expect(rendered.container).toMatchSnapshot();
+  });
+
+  describe("when extension navigates to route without parameters", () => {
+    beforeEach(() => {
+      testExtension.navigate();
+    });
+
+    it("renders", () => {
+      expect(rendered.container).toMatchSnapshot();
+    });
+
+    it("URL is correct", () => {
+      expect(currentPath.get()).toBe("/extension/some-extension-id");
+    });
+
+    it("query parameters is empty", () => {
+      expect(queryParameters.get()).toEqual({});
+    });
+
+    describe("when changing page parameters", () => {
+      beforeEach(() => {
+        const button = rendered.getByTestId("button-to-change-page-parameters");
```

Comment:
I think getByTestId should be avoided and getByText etc. should be preferred.

See best practise in the official docs: https://testing-library.com/docs/queries/bytestid/

Reply:
Read the post linked in the page you provided. https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change

Reply:
I read it, and it agrees with my review comment..

To be more specific: using test ids is okay if other queries are not reliable / possible.

Quote from the post:
"Sometimes you can't reliably select an element by any of the other queries. For those, it's recommended to use data-testid"

Longer quote from the post:
"This is why [Testing Library](https://testing-library.com/) has the queries that it does. The queries help you to find elements in the same way that users will find them. These queries allow you to find elements by their [role](https://testing-library.com/docs/dom-testing-library/api-queries#byrole), [label](https://testing-library.com/docs/dom-testing-library/api-queries#bylabeltext), [placeholder](https://testing-library.com/docs/dom-testing-library/api-queries#byplaceholdertext), [text contents](https://testing-library.com/docs/dom-testing-library/api-queries#bytext), [display value](https://testing-library.com/docs/dom-testing-library/api-queries#bydisplayvalue), [alt text](https://testing-library.com/docs/dom-testing-library/api-queries#byalttext), [title](https://testing-library.com/docs/dom-testing-library/api-queries#bytitle), [test ID](https://testing-library.com/docs/dom-testing-library/api-queries#bytestid).

That's actually in the order of [recommendation](https://testing-library.com/docs/guide-which-query)."

The priority is even documented here:
https://testing-library.com/docs/queries/about/#priority

That's why I said that others should be preferred.

I'm fine with using test ids if others are not possible / practical for some reason.

Reply:
There is no way to reliably select an element using `title`, `role`, `placeholder`, `text contents`, `display value`, `alt text` or `title`. Role is only one of those that is not changed too often. But role alone isn't enough. Therefore, only option is to select with `test ID`. 

Needs to be said again, I'm not interested in the tools and how we use them. Tool (here. testing-library) doesn't affect my mind in anyway. I'm more interested about the problem that we are trying to solve. In this case it is: 

**How do we test something that user can act with in a way that:**
1. we can minimize amount of unit tests that need to adapt for unrelated change
2. we can change our mind in the content
3. we can even change our mind on the element (e.g. could be button now, but in the future we decide that it's a link)
4. we can be sure enough that user find's the element (doesn't have to be same way, since it's just a burden. As long as abstraction levels are correct in the unit tests, we should be fine)

Reply:
Clearly it's possible to reliably select an element using title, label, role, placeholder, text contents, display value or alt text.

Usually it's good to follow library's best practises.

Reply:
How? Could you elaborate how you would solve the problem that I introduced in previous message?